boot/seal.go

```diff
@@ -25,65 +25,269 @@ import (
-
-	// the path to the run mode grub EFI binary
-	runbf := bootloader.NewBootFile("", filepath.Join(InitramfsUbuntuBootDir, "EFI/boot/grubx64.efi"), bootloader.RoleRunMode)
-	loadChain = secboot.NewLoadChain(runbf, loadChain)
-	// the path to the recovery grub EFI binary
-	recbf := bootloader.NewBootFile("", filepath.Join(InitramfsUbuntuSeedDir, "EFI/boot/grubx64.efi"), bootloader.RoleRecovery)
-	loadChain = secboot.NewLoadChain(recbf, loadChain)
-	// the path to the shim EFI binary
-	shimbf := bootloader.NewBootFile("", filepath.Join(InitramfsUbuntuSeedDir, "EFI/boot/bootx64.efi"), bootloader.RoleRecovery)
-	loadChain = secboot.NewLoadChain(shimbf, loadChain)
-
-	// Get the expected kernel command line for the system that is currently being installed
+	// build the recovery mode boot chain
+	rbl, err := bootloader.Find(InitramfsUbuntuSeedDir, &bootloader.Options{
+		Role: bootloader.RoleRecovery,
+	})
+	if err != nil {
+		return fmt.Errorf("cannot find the recovery bootloader: %v", err)
+	}
+
+	recoveryBootChain, err := buildRecoveryBootChain(rbl, model, modeenv)
+	if err != nil {
+		return fmt.Errorf("cannot build recovery boot chain: %v", err)
+	}
+
+	// build the run mode boot chains
+	bl, err := bootloader.Find(InitramfsUbuntuBootDir, &bootloader.Options{
+		Role:        bootloader.RoleRunMode,
+		NoSlashBoot: true,
+	})
+	if err != nil {
+		return fmt.Errorf("cannot find the bootloader: %v", err)
+	}
 	cmdline, err := ComposeCandidateCommandLine(model)
 	if err != nil {
-		return fmt.Errorf("cannot obtain kernel command line: %v", err)
+		return fmt.Errorf("cannot compose the candidate command line: %v", err)
 	}
 
-	// Get the expected kernel command line of the recovery system we're installing from
-	recoveryCmdline, err := ComposeRecoveryCommandLine(model, modeenv.RecoverySystem)
+	runModeBootChains, err := buildRunModeBootChains(rbl, bl, model, modeenv, cmdline)
 	if err != nil {
-		return fmt.Errorf("cannot obtain recovery kernel command line: %v", err)
+		return fmt.Errorf("cannot build run mode boot chain: %v", err)
 	}
 
-	kernelCmdlines := []string{
-		cmdline,
-		recoveryCmdline,
+	pbc := toPredictableBootChains(append(runModeBootChains, recoveryBootChain))
+
+	// XXX: store the predictable bootchains
+
+	roleToBlName := map[bootloader.Role]string{
+		bootloader.RoleRecovery: rbl.Name(),
+		bootloader.RoleRunMode:  bl.Name(),
 	}
 
-	sealKeyParams := secboot.SealKeyParams{
-		ModelParams: []*secboot.SealKeyModelParams{
-			{
-				Model:          model,
-				KernelCmdlines: kernelCmdlines,
-				EFILoadChains:  []*secboot.LoadChain{loadChain},
-			},
-		},
-		KeyFile:                 filepath.Join(InitramfsEncryptionKeyDir, "ubuntu-data.sealed-key"),
-		TPMPolicyUpdateDataFile: filepath.Join(InstallHostFDEDataDir, "policy-update-data"),
-		TPMLockoutAuthFile:      filepath.Join(InstallHostFDEDataDir, "tpm-lockout-auth"),
+	// get parameters from bootchains and seal the key
+	params, err := sealKeyParams(pbc, roleToBlName)
```

Comment:
I'm happy to chat, but I don't understand the comment. I asked it to take the ordered ones, because we can do some easy coalescing with the ordered ones (which tbh the current code here is not doing)

Reply:
Ok, let me look at the branch locally, I'm probably missing something there.

Reply:
Discussed in a chat, the problem is with generating predictable boot chains, the boot assets inside a chain should not be reodered. I'll push a PR for that.

Reply:
Opened #9313